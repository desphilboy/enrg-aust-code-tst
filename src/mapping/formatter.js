import { DEFAULT_RECORDER } from './constants.js';
import _ from 'lodash';

export const flattenData = (festivalData) =>
	(festivalData || []).reduce((acc, fd) => {
		fd.bands?.forEach((band) => {
			acc = [
				...acc,
				{
					recordLabel: band.recordLabel || DEFAULT_RECORDER,
					bandName: band.name,
					festival: fd.name,
				},
			];
		});
		return acc;
	}, []);

export const sortRls = (a, b) => {
	if (a.recordLabel === DEFAULT_RECORDER) {
		return 1;
	}
	if (b.recordLabel === DEFAULT_RECORDER) {
		return -1;
	}
	return a.recordLabel > b.recordLabel ? 1 : -1;
};

export const groupedRecordLabel = (flatData) =>
	_.chain(flatData)
		// Group the elements of Array based on recordlabel
		.groupBy('recordLabel')
		.map((value, key) => ({ recordLabel: key, bands: value }))
		.value()
		.sort(sortRls);

export const makePrintable = (grpData) =>
	grpData.map(({ recordLabel, bands }) => ({
		recordLabel,
		bands: _.chain(bands)
			.groupBy('bandName')
			.map((value, key) => ({ band: key, festivals: value }))
			.value()
			.sort((a, b) => {
				return a.bandName > b.bandName ? 1 : -1;
			}),
	}));
