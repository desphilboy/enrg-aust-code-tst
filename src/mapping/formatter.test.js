import { flattenData, groupedRecordLabel, makePrintable } from './formatter.js';
import { SAMPLE } from './__fixture__/sample_festivals.js';

describe('Formatter', () => {
	describe('flatten data', () => {
		it('formats data according to spec', () => {
			const flatData = flattenData(SAMPLE);

			expect(flatData[0].bandName).toBe('Squint-281');
			expect(flatData).toMatchSnapshot();
		});
	});

	describe('groupRecordLabel', () => {
		it('groups data based on record labelaccording to spec', () => {
			const flatData = flattenData(SAMPLE);
			const grouped = groupedRecordLabel(flatData);
			expect(grouped).toMatchSnapshot();
			expect(grouped[0]).toEqual({
				bands: [
					{
						bandName: 'Manish Ditch',
						festival: 'Trainerella',
						recordLabel: 'ACR',
					},
					{
						bandName: 'Critter Girls',
						festival: undefined,
						recordLabel: 'ACR',
					},
				],
				recordLabel: 'ACR',
			});
			expect(Array.isArray(grouped[0].bands)).toBe(true);
		});

		it('sorts based on recordLabel', () => {
			const flatData = flattenData(SAMPLE);
			const grouped = groupedRecordLabel(flatData);
			expect(grouped.map((g) => g.recordLabel)).toEqual([
				'ACR',
				'Anti Records',
				'Fourth Woman Records',
				'MEDIOCRE Music',
				'Marner Sis. Recording',
				'Monocracy Records',
				'Outerscope',
				'Pacific Records',
				'Still Bottom Records',
				'XS Recordings',
				'Unknown recorder',
			]);
		});
	});
	describe('makePrintable', () => {
		it('matches the requested output', () => {
			const flatData = flattenData(SAMPLE);
			const grouped = groupedRecordLabel(flatData);
			const printable = makePrintable(grouped);

			expect(printable).toMatchSnapshot();
		});

		it('sorts bands by bandname', () => {
			const flatData = flattenData(SAMPLE);
			const grouped = groupedRecordLabel(flatData);
			const printable = makePrintable(grouped);

			expect(printable[0].bands.map((p) => p.band)).toEqual([
				'Critter Girls',
				'Manish Ditch',
			]);

			expect(printable[1].bands.map((p) => p.band)).toEqual(['YOUKRANE']);
		});
	});
});
