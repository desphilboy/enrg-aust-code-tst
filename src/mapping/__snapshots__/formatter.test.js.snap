// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Formatter flatten data formats data according to spec 1`] = `
Array [
  Object {
    "bandName": "Squint-281",
    "festival": "Small Night In",
    "recordLabel": "Outerscope",
  },
  Object {
    "bandName": "The Black Dashes",
    "festival": "Small Night In",
    "recordLabel": "Fourth Woman Records",
  },
  Object {
    "bandName": "Green Mild Cold Capsicum",
    "festival": "Small Night In",
    "recordLabel": "Marner Sis. Recording",
  },
  Object {
    "bandName": "Yanke East",
    "festival": "Small Night In",
    "recordLabel": "MEDIOCRE Music",
  },
  Object {
    "bandName": "Wild Antelope",
    "festival": "Small Night In",
    "recordLabel": "Marner Sis. Recording",
  },
  Object {
    "bandName": "Squint-281",
    "festival": "Trainerella",
    "recordLabel": "Outerscope",
  },
  Object {
    "bandName": "Wild Antelope",
    "festival": "Trainerella",
    "recordLabel": "Still Bottom Records",
  },
  Object {
    "bandName": "Manish Ditch",
    "festival": "Trainerella",
    "recordLabel": "ACR",
  },
  Object {
    "bandName": "Adrian Venti",
    "festival": "Trainerella",
    "recordLabel": "Monocracy Records",
  },
  Object {
    "bandName": "YOUKRANE",
    "festival": "Trainerella",
    "recordLabel": "Anti Records",
  },
  Object {
    "bandName": "Critter Girls",
    "festival": undefined,
    "recordLabel": "ACR",
  },
  Object {
    "bandName": "Propeller",
    "festival": undefined,
    "recordLabel": "Pacific Records",
  },
  Object {
    "bandName": "Winter Primates",
    "festival": "LOL-palooza",
    "recordLabel": "Unknown recorder",
  },
  Object {
    "bandName": "Frank Jupiter",
    "festival": "LOL-palooza",
    "recordLabel": "Pacific Records",
  },
  Object {
    "bandName": "Jill Black",
    "festival": "LOL-palooza",
    "recordLabel": "Fourth Woman Records",
  },
  Object {
    "bandName": "Werewolf Weekday",
    "festival": "LOL-palooza",
    "recordLabel": "XS Recordings",
  },
  Object {
    "bandName": "Summon",
    "festival": "Twisted Tour",
    "recordLabel": "Outerscope",
  },
  Object {
    "bandName": "Auditones",
    "festival": "Twisted Tour",
    "recordLabel": "Marner Sis. Recording",
  },
  Object {
    "bandName": "Squint-281",
    "festival": "Twisted Tour",
    "recordLabel": "Unknown recorder",
  },
]
`;

exports[`Formatter groupRecordLabel groups data based on record labelaccording to spec 1`] = `
Array [
  Object {
    "bands": Array [
      Object {
        "bandName": "Manish Ditch",
        "festival": "Trainerella",
        "recordLabel": "ACR",
      },
      Object {
        "bandName": "Critter Girls",
        "festival": undefined,
        "recordLabel": "ACR",
      },
    ],
    "recordLabel": "ACR",
  },
  Object {
    "bands": Array [
      Object {
        "bandName": "YOUKRANE",
        "festival": "Trainerella",
        "recordLabel": "Anti Records",
      },
    ],
    "recordLabel": "Anti Records",
  },
  Object {
    "bands": Array [
      Object {
        "bandName": "The Black Dashes",
        "festival": "Small Night In",
        "recordLabel": "Fourth Woman Records",
      },
      Object {
        "bandName": "Jill Black",
        "festival": "LOL-palooza",
        "recordLabel": "Fourth Woman Records",
      },
    ],
    "recordLabel": "Fourth Woman Records",
  },
  Object {
    "bands": Array [
      Object {
        "bandName": "Yanke East",
        "festival": "Small Night In",
        "recordLabel": "MEDIOCRE Music",
      },
    ],
    "recordLabel": "MEDIOCRE Music",
  },
  Object {
    "bands": Array [
      Object {
        "bandName": "Green Mild Cold Capsicum",
        "festival": "Small Night In",
        "recordLabel": "Marner Sis. Recording",
      },
      Object {
        "bandName": "Wild Antelope",
        "festival": "Small Night In",
        "recordLabel": "Marner Sis. Recording",
      },
      Object {
        "bandName": "Auditones",
        "festival": "Twisted Tour",
        "recordLabel": "Marner Sis. Recording",
      },
    ],
    "recordLabel": "Marner Sis. Recording",
  },
  Object {
    "bands": Array [
      Object {
        "bandName": "Adrian Venti",
        "festival": "Trainerella",
        "recordLabel": "Monocracy Records",
      },
    ],
    "recordLabel": "Monocracy Records",
  },
  Object {
    "bands": Array [
      Object {
        "bandName": "Squint-281",
        "festival": "Small Night In",
        "recordLabel": "Outerscope",
      },
      Object {
        "bandName": "Squint-281",
        "festival": "Trainerella",
        "recordLabel": "Outerscope",
      },
      Object {
        "bandName": "Summon",
        "festival": "Twisted Tour",
        "recordLabel": "Outerscope",
      },
    ],
    "recordLabel": "Outerscope",
  },
  Object {
    "bands": Array [
      Object {
        "bandName": "Propeller",
        "festival": undefined,
        "recordLabel": "Pacific Records",
      },
      Object {
        "bandName": "Frank Jupiter",
        "festival": "LOL-palooza",
        "recordLabel": "Pacific Records",
      },
    ],
    "recordLabel": "Pacific Records",
  },
  Object {
    "bands": Array [
      Object {
        "bandName": "Wild Antelope",
        "festival": "Trainerella",
        "recordLabel": "Still Bottom Records",
      },
    ],
    "recordLabel": "Still Bottom Records",
  },
  Object {
    "bands": Array [
      Object {
        "bandName": "Werewolf Weekday",
        "festival": "LOL-palooza",
        "recordLabel": "XS Recordings",
      },
    ],
    "recordLabel": "XS Recordings",
  },
  Object {
    "bands": Array [
      Object {
        "bandName": "Winter Primates",
        "festival": "LOL-palooza",
        "recordLabel": "Unknown recorder",
      },
      Object {
        "bandName": "Squint-281",
        "festival": "Twisted Tour",
        "recordLabel": "Unknown recorder",
      },
    ],
    "recordLabel": "Unknown recorder",
  },
]
`;

exports[`Formatter makePrintable matches the requested output 1`] = `
Array [
  Object {
    "bands": Array [
      Object {
        "band": "Critter Girls",
        "festivals": Array [
          Object {
            "bandName": "Critter Girls",
            "festival": undefined,
            "recordLabel": "ACR",
          },
        ],
      },
      Object {
        "band": "Manish Ditch",
        "festivals": Array [
          Object {
            "bandName": "Manish Ditch",
            "festival": "Trainerella",
            "recordLabel": "ACR",
          },
        ],
      },
    ],
    "recordLabel": "ACR",
  },
  Object {
    "bands": Array [
      Object {
        "band": "YOUKRANE",
        "festivals": Array [
          Object {
            "bandName": "YOUKRANE",
            "festival": "Trainerella",
            "recordLabel": "Anti Records",
          },
        ],
      },
    ],
    "recordLabel": "Anti Records",
  },
  Object {
    "bands": Array [
      Object {
        "band": "Jill Black",
        "festivals": Array [
          Object {
            "bandName": "Jill Black",
            "festival": "LOL-palooza",
            "recordLabel": "Fourth Woman Records",
          },
        ],
      },
      Object {
        "band": "The Black Dashes",
        "festivals": Array [
          Object {
            "bandName": "The Black Dashes",
            "festival": "Small Night In",
            "recordLabel": "Fourth Woman Records",
          },
        ],
      },
    ],
    "recordLabel": "Fourth Woman Records",
  },
  Object {
    "bands": Array [
      Object {
        "band": "Yanke East",
        "festivals": Array [
          Object {
            "bandName": "Yanke East",
            "festival": "Small Night In",
            "recordLabel": "MEDIOCRE Music",
          },
        ],
      },
    ],
    "recordLabel": "MEDIOCRE Music",
  },
  Object {
    "bands": Array [
      Object {
        "band": "Auditones",
        "festivals": Array [
          Object {
            "bandName": "Auditones",
            "festival": "Twisted Tour",
            "recordLabel": "Marner Sis. Recording",
          },
        ],
      },
      Object {
        "band": "Wild Antelope",
        "festivals": Array [
          Object {
            "bandName": "Wild Antelope",
            "festival": "Small Night In",
            "recordLabel": "Marner Sis. Recording",
          },
        ],
      },
      Object {
        "band": "Green Mild Cold Capsicum",
        "festivals": Array [
          Object {
            "bandName": "Green Mild Cold Capsicum",
            "festival": "Small Night In",
            "recordLabel": "Marner Sis. Recording",
          },
        ],
      },
    ],
    "recordLabel": "Marner Sis. Recording",
  },
  Object {
    "bands": Array [
      Object {
        "band": "Adrian Venti",
        "festivals": Array [
          Object {
            "bandName": "Adrian Venti",
            "festival": "Trainerella",
            "recordLabel": "Monocracy Records",
          },
        ],
      },
    ],
    "recordLabel": "Monocracy Records",
  },
  Object {
    "bands": Array [
      Object {
        "band": "Summon",
        "festivals": Array [
          Object {
            "bandName": "Summon",
            "festival": "Twisted Tour",
            "recordLabel": "Outerscope",
          },
        ],
      },
      Object {
        "band": "Squint-281",
        "festivals": Array [
          Object {
            "bandName": "Squint-281",
            "festival": "Small Night In",
            "recordLabel": "Outerscope",
          },
          Object {
            "bandName": "Squint-281",
            "festival": "Trainerella",
            "recordLabel": "Outerscope",
          },
        ],
      },
    ],
    "recordLabel": "Outerscope",
  },
  Object {
    "bands": Array [
      Object {
        "band": "Frank Jupiter",
        "festivals": Array [
          Object {
            "bandName": "Frank Jupiter",
            "festival": "LOL-palooza",
            "recordLabel": "Pacific Records",
          },
        ],
      },
      Object {
        "band": "Propeller",
        "festivals": Array [
          Object {
            "bandName": "Propeller",
            "festival": undefined,
            "recordLabel": "Pacific Records",
          },
        ],
      },
    ],
    "recordLabel": "Pacific Records",
  },
  Object {
    "bands": Array [
      Object {
        "band": "Wild Antelope",
        "festivals": Array [
          Object {
            "bandName": "Wild Antelope",
            "festival": "Trainerella",
            "recordLabel": "Still Bottom Records",
          },
        ],
      },
    ],
    "recordLabel": "Still Bottom Records",
  },
  Object {
    "bands": Array [
      Object {
        "band": "Werewolf Weekday",
        "festivals": Array [
          Object {
            "bandName": "Werewolf Weekday",
            "festival": "LOL-palooza",
            "recordLabel": "XS Recordings",
          },
        ],
      },
    ],
    "recordLabel": "XS Recordings",
  },
  Object {
    "bands": Array [
      Object {
        "band": "Squint-281",
        "festivals": Array [
          Object {
            "bandName": "Squint-281",
            "festival": "Twisted Tour",
            "recordLabel": "Unknown recorder",
          },
        ],
      },
      Object {
        "band": "Winter Primates",
        "festivals": Array [
          Object {
            "bandName": "Winter Primates",
            "festival": "LOL-palooza",
            "recordLabel": "Unknown recorder",
          },
        ],
      },
    ],
    "recordLabel": "Unknown recorder",
  },
]
`;
